rust/backend/src/configuration.rs

```diff
@@ -0,0 +1,87 @@
+// SPDX-FileCopyrightText: 2024 Benedikt Zinn <benedikt.wh.zinn@gmail.com>
+//
+// SPDX-License-Identifier: MIT
+
+use shared::config::{Configuration as ProtoConfig, EbpfEntry as ProtoEbpfEntry};
+use std::fs::File;
+use std::io;
+use std::io::{BufReader, BufWriter};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct Configuration {
+    entries: Vec<EbpfEntry>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+struct EbpfEntry {
+    hr_name: String,            // human-readable name, for display on the frontend
+    description: String,        // human-readable description of the probe
+    ebpf_name: String,          // name of the ebpf function to actually load
+    fn_id: u64,                 // id of the daemon function to handle
+}
+
+impl From<Configuration> for ProtoConfig {
+    fn from(config: Configuration) -> Self {
+        ProtoConfig {
+            entries: config
+                .entries
+                .into_iter()
+                .map(ProtoEbpfEntry::from)
+                .collect(),
+        }
+    }
+}
+
+impl From<ProtoConfig> for Configuration {
+    fn from(proto: ProtoConfig) -> Self {
+        Configuration {
+            entries: proto.entries.into_iter().map(EbpfEntry::from).collect(),
+        }
+    }
+}
+
+impl From<EbpfEntry> for ProtoEbpfEntry {
+    fn from(entry: EbpfEntry) -> Self {
+        ProtoEbpfEntry {
+            hr_name: entry.hr_name,
+            description: entry.description,
+            ebpf_name: entry.ebpf_name,
+            fn_id: entry.fn_id,
+        }
+    }
+}
+
+impl From<ProtoEbpfEntry> for EbpfEntry {
+    fn from(proto: ProtoEbpfEntry) -> Self {
+        EbpfEntry {
+            hr_name: proto.hr_name,
+            description: proto.description,
+            ebpf_name: proto.ebpf_name,
+            fn_id: proto.fn_id,
+        }
+    }
+}
+
+impl Configuration {
+    pub fn load_from_file(path: &str) -> io::Result<Self> {
+        let file = File::open(path)?;
+        let reader = BufReader::new(file);
+        let config = serde_json::from_reader(reader)?;
+        Ok(config)
+    }
+
+    pub fn save_to_file(&self, path: &str) -> io::Result<()> {
+        let file = File::create(path)?;
+        let writer = BufWriter::new(file);
+        serde_json::to_writer(writer, &self)?;
+        Ok(())
+    }
+
+    pub fn validate(&self) -> Result<(), io::Error> {
+        //TODO: Implement this function
+        Ok(())
+    }
+}
```

Comment:
In general, I would replace that with the following in the `build.rs` of the protobuf library (`shared`):

```
tonic_build::configure()
        .protoc_arg("--experimental_allow_proto3_optional")
        .type_attribute(".", "#[derive(Serialize, Deserialize)]")
        .compile_protos(&[
            "../proto/counter.proto",
            "../proto/ziofa.proto",
            "../proto/config.proto"
        ], &["../proto"])
        .unwrap();
```

Then you share the type definition for both serde and protobuf.

For more information see https://docs.rs/tonic-build/0.6.0/tonic_build/struct.Builder.html#method.type_attribute and https://docs.rs/prost-build/latest/prost_build/struct.Config.html#method.type_attribute